californium-core/src/test/java/org/eclipse/californium/core/network/UdpMatcherTest.java

```diff
@@ -62,6 +65,28 @@ public void testReceiveResponseAcceptsResponseWithoutCorrelationInformation() {
 	}
 
 	@Test
+	public void testCancelledRequestShouldRemoveFromTokenMap() {
+		UdpMatcher matcher = newMatcher(false);
+		Exchange exchange = sendRequest(matcher, null);
+
+		exchange.getRequest().cancel();
+
+		Exchange matched = matcher.receiveResponse(responseFor(exchange.getCurrentRequest()), null);
+		assertNull(matched);
+	}
+
+	@Test
+	public void testCancelledRequestShouldRemoveFromMidMap() {
```

Comment:
I don't think that this test is relevant for the kind of requests that cause the leaking:
* CON
* separately ACKed (i.e. already removed from the MID map because the ACK contained the original request's MID)
* no response received yet